src/sync/impl/sync_client.hpp

```diff
@@ -33,12 +45,17 @@ struct SyncClient {
 
     SyncClient(std::unique_ptr<util::Logger> logger,
                ReconnectMode reconnect_mode = ReconnectMode::normal,
-               bool verify_ssl = true)
-    : client(make_client(*logger, reconnect_mode, verify_ssl)) // Throws
-    , m_logger(std::move(logger))
-    , m_thread([this] {
-        client.run();
-    }) // Throws
+               bool verify_ssl = true,
+               ClientThreadListener *client_thread_listener = nullptr)
+            : client(make_client(*logger, reconnect_mode, verify_ssl)) // Throws
+            , m_logger(std::move(logger)),
+              m_thread([this, client_thread_listener] {
+                  if (client_thread_listener)
+                      client_thread_listener->on_client_thread_ready();
+                  client.run();
+                  if (client_thread_listener)
+                      client_thread_listener->on_client_thread_closing();
+              }) // Throws
```

Comment:
This indentation is still very weird. It should look something like:
```c++
SyncClient(std::unique_ptr<util::Logger> logger,
           ReconnectMode reconnect_mode = ReconnectMode::normal,
           bool verify_ssl = true,
           ClientThreadListener *client_thread_listener = nullptr)
: client(make_client(*logger, reconnect_mode, verify_ssl)) // Throws
, m_logger(std::move(logger)),
  m_thread([this, client_thread_listener] {
    if (client_thread_listener)
        client_thread_listener->on_client_thread_ready();
    client.run();
    if (client_thread_listener)
        client_thread_listener->on_client_thread_closing();
}) // Throws
{
}
```

This makes me notice that the calls to the client thread listener are not exception safe. If `client.run()` throws an exception, which it can, then the thread will be closed but `on_client_thread_closing` will never be called.